File: flutter/lib/core/widgets/connection_status_banner.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../data/models/connection_status.dart';
import '../../features/sessions/providers/sessions_provider.dart';

class ConnectionStatusBanner extends ConsumerStatefulWidget {
  const ConnectionStatusBanner({super.key});

  @override
  ConsumerState<ConnectionStatusBanner> createState() =>
      _ConnectionStatusBannerState();
}

class _ConnectionStatusBannerState
    extends ConsumerState<ConnectionStatusBanner> {
  bool _visible = false;
  ConnectionStatus? _status;
  Timer? _hideTimer;
  Timer? _countdownTimer;
  int _countdown = 5;

  @override
  void dispose() {
    _hideTimer?.cancel();
    _countdownTimer?.cancel();
    super.dispose();
  }

  void _onStatusChanged(ConnectionStatus? prev, ConnectionStatus next) {
    if (next == _status) return;

    _hideTimer?.cancel();
    _countdownTimer?.cancel();

    if (next == ConnectionStatus.connected) {
      setState(() {
        _status = next;
        _visible = true;
      });
      _hideTimer = Timer(const Duration(seconds: 2), () {
        if (mounted) setState(() => _visible = false);
      });
    } else if (next == ConnectionStatus.offline) {
      setState(() {
        _status = next;
        _visible = true;
        _countdown = 5;
      });
      _countdownTimer = Timer.periodic(const Duration(seconds: 1), (_) {
        if (!mounted) return;
        setState(() {
          _countdown = _countdown > 0 ? _countdown - 1 : 5;
        });
      });
    } else {
      setState(() {
        _status = next;
        _visible = true;
        _countdown = 5;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    ref.listen<AsyncValue<ConnectionStatus>>(
      connectionStatusProvider,
      (prev, next) {
        next.whenData((status) =>
            _onStatusChanged(prev?.valueOrNull, status));
      },
    );

    final bool show = _visible && _status != null;

    Widget content;
    Color bgColor;

    final status = _status;
    if (status == ConnectionStatus.connected) {
      bgColor = Colors.green.shade700;
      content = const Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Icon(Icons.check_circle_outline, color: Colors.white, size: 14),
          SizedBox(width: 6),
          Text(
            'Connected',
            style: TextStyle(
              color: Colors.white,
              fontSize: 12,
              fontWeight: FontWeight.w500,
            ),
          ),
        ],
      );
    } else if (status == ConnectionStatus.offline) {
      bgColor = Colors.red.shade700;
      content = Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          const Icon(Icons.wifi_off, color: Colors.white, size: 14),
          const SizedBox(width: 6),
          Text(
            'Server unreachable — retrying in ${_countdown}s',
            style: const TextStyle(
              color: Colors.white,
              fontSize: 12,
              fontWeight: FontWeight.w500,
            ),
          ),
        ],
      );
    } else {
      bgColor = Colors.orange.shade700;
      final label = status == ConnectionStatus.reconnecting
          ? 'Reconnecting...'
          : 'Connecting...';
      content = Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          const SizedBox(
            width: 12,
            height: 12,
            child: CircularProgressIndicator(
              strokeWidth: 2,
              valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
            ),
          ),
          const SizedBox(width: 8),
          Text(
            label,
            style: const TextStyle(
              color: Colors.white,
              fontSize: 12,
              fontWeight: FontWeight.w500,
            ),
          ),
        ],
      );
    }

    return AnimatedContainer(
      duration: const Duration(milliseconds: 300),
      height: show ? 36 : 0,
      child: show
          ? AnimatedSwitcher(
              duration: const Duration(milliseconds: 200),
              child: Container(
                key: ValueKey(status),
                width: double.infinity,
                color: bgColor,
                alignment: Alignment.center,
                padding: const EdgeInsets.symmetric(horizontal: 16),
                child: content,
              ),
            )
          : const SizedBox.shrink(),
    );
  }
}

File: android-native/app/src/main/java/com/agentshell/core/widgets/ConnectionStatusBanner.kt
package com.agentshell.core.widgets

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.expandVertically
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.WifiOff
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.agentshell.data.model.ConnectionStatus
import kotlinx.coroutines.delay

/**
 * Animated status banner that mirrors Flutter's ConnectionStatusBanner.
 *
 * CONNECTED  → green banner visible for 2 s then fades away
 * CONNECTING → orange banner "Connecting…"
 * RECONNECTING → orange banner "Reconnecting…"
 * OFFLINE    → red banner with countdown to next retry
 *
 * Caller provides [connectionStatus]; the composable manages its own
 * visibility state so the banner auto-dismisses after a connected flash.
 */
@Composable
fun ConnectionStatusBanner(
    connectionStatus: ConnectionStatus,
) {
    var visible by remember { mutableStateOf(false) }
    var displayedStatus by remember { mutableStateOf(connectionStatus) }
    var countdown by remember { mutableIntStateOf(5) }

    // React to status changes
    LaunchedEffect(connectionStatus) {
        displayedStatus = connectionStatus
        when (connectionStatus) {
            ConnectionStatus.CONNECTED -> {
                visible = true
                // Auto-hide after 2 s on connected
                delay(2_000)
                visible = false
            }
            ConnectionStatus.OFFLINE -> {
                visible = true
                // Tick countdown
                countdown = 5
                while (true) {
                    delay(1_000)
                    countdown = if (countdown > 0) countdown - 1 else 5
                }
            }
            else -> {
                visible = true
                countdown = 5
            }
        }
    }

    AnimatedVisibility(
        visible = visible,
        enter = expandVertically(expandFrom = androidx.compose.ui.Alignment.Top),
        exit = shrinkVertically(shrinkTowards = androidx.compose.ui.Alignment.Top),
    ) {
        val (bgColor, content) = when (displayedStatus) {
            ConnectionStatus.CONNECTED -> Pair(
                Color(0xFF388E3C),
                @Composable {
                    Icon(
                        Icons.Default.CheckCircle,
                        contentDescription = null,
                        tint = Color.White,
                        modifier = Modifier.size(14.dp),
                    )
                    Spacer(Modifier.width(6.dp))
                    Text(
                        "Connected",
                        color = Color.White,
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Medium,
                    )
                },
            )
            ConnectionStatus.OFFLINE -> Pair(
                Color(0xFFC62828),
                @Composable {
                    Icon(
                        Icons.Default.WifiOff,
                        contentDescription = null,
                        tint = Color.White,
                        modifier = Modifier.size(14.dp),
                    )
                    Spacer(Modifier.width(6.dp))
                    Text(
                        "Server unreachable — retrying in ${countdown}s",
                        color = Color.White,
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Medium,
                    )
                },
            )
            ConnectionStatus.RECONNECTING -> Pair(
                Color(0xFFE65100),
                @Composable {
                    CircularProgressIndicator(
                        modifier = Modifier.size(12.dp),
                        color = Color.White,
                        strokeWidth = 2.dp,
                    )
                    Spacer(Modifier.width(8.dp))
                    Text(
                        "Reconnecting...",
                        color = Color.White,
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Medium,
                    )
                },
            )
            ConnectionStatus.CONNECTING -> Pair(
                Color(0xFFE65100),
                @Composable {
                    CircularProgressIndicator(
                        modifier = Modifier.size(12.dp),
                        color = Color.White,
                        strokeWidth = 2.dp,
                    )
                    Spacer(Modifier.width(8.dp))
                    Text(
                        "Connecting...",
                        color = Color.White,
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Medium,
                    )
                },
            )
        }

        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(36.dp)
                .background(bgColor)
                .padding(horizontal = 16.dp),
            contentAlignment = Alignment.Center,
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
            ) {
                content()
            }
        }
    }
}
